doc/userguide/devguide/internals/engines/stream/inspection_raw_data.rst

```diff
@@ -0,0 +1,117 @@
+Inspection of raw stream data
+#############################
+
+Stream Engine
+^^^^^^^^^^^^^
+
+Suricata's Stream engine tracks and processes all the TCP stream data. Its responsibilities include
+
+* TCP segment reassembly
+* TCP data normalization
+* gap management and handling
+* maintaining internal caches
+* handling of special cases like TCP URG ptr
+* applying user-defined constraints like stream depth etc
+
+for IDS as well as inline mode.
+
+Internal storage of stream data
+===============================
+
+For a stream with small gaps, a Red Black Tree is used to store the streaming buffer blocks.
+
+For a stream with large gaps, regions (list of blocks of data) are used.
+
+For a stream without gaps, one continuos streaming buffer is used (i.e. just one region).
+
+These different data structures are used in a quest to make efficient use of memory in exceptional
+conditions.
+
+Role of stream reassembly
+=========================
+
+TCP stream data can arrive in any manner. For example, 100 bytes of data can arrive as 100 bytes at
+once or 1 byte at a time in 100 segments. The possibilities are insanely high! So, if it's
+100 bytes of data, there are :math:`2^{99}` ways this data can be received in a world where this
+data arrives in order!
+
+Hence, it is important for the engine to reassemble the TCP stream data to avoid unnecessary
+inspection on incomplete data and to avoid leaving room for evasion techniques based on small
+segments. Stream reassembly makes sure that the data to be matched upon is reliable.
+
+Role of Detection Engine and Applayer Parser
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to conduct inspection on certain stream data, the detection engine has to request the
```

Comment:
also we have https://redmine.openinfosecfoundation.org/issues/7328

Reply:
Super cool. Is this in the works?